Document CI/CD (#12980)
PyTorchLightning/pytorch-lightning: .github/workflows/README.md

| Test slow                   | .github/workflows/ci_test-slow.yml  | Run only slow tests. Slow tests usually need to spawn threads and cannot be speed up or simplified.                                                            | CPU           | (3.7, 1.8)                                       | linux, mac, windows |
| PL.pytorch-lightning (IPUs) | .azure-pipelines/ipu-tests.yml      | Run only IPU-specific tests.                                                                                                                                   | IPU           | (3.8, 1.9)                                       | linux               |
| PL.pytorch-lightning (HPUs) | .azure-pipelines/hpu-tests.yml      | Run only HPU-specific tests.                                                                                                                                   | HPU           | (3.8, 1.10)                                      | linux               |

> | PL.pytorch-lightning (GPUs) | .azure-pipelines/gpu-tests.yml      | Run all CPU and GPU-specific tests, standalone and examples. Each standalone test needs to run individually to avoid unwanted interactions between test cases. | GPU           | (3.7, 1.8)                                       | linux               |

Author: ```suggestion
| PL.pytorch-lightning (GPUs) | .azure-pipelines/gpu-tests.yml      | Run all CPU and GPU-specific tests, standalone and examples. Each standalone test needs to be run in separate processes to avoid unwanted interactions between test cases. | GPU           | (3.7, 1.8)                                       | linux               |
```

| PyTorchLightning.Benchmark  | .azure-pipelines/gpu-benchmark.yml  | Run speed/memory benchmarks for parity with pure PyTorch.                                                                                                                  | GPU           | (3.7, 1.8)                                       | linux               |
| test-on-tpus                | .circleci/config.yml                | Run only TPU-specific tests.                                                                                                                                               | TPU           | (3.7, 1.9)                                       | linux               |

- \*Accelerators used in CI
  - GPU: 2 x NVIDIA P100

>   - TPU: Google GKE TPUv2/3

Reviewer: ```suggestion
  - TPU: Google GKE TPUv2
```
@kaushikb11 atm we use only v2, correct?

Reviewer 2: Nope, it's TPU v3 https://github.com/PyTorchLightning/pytorch-lightning/blob/master/dockers/tpu-tests/tpu_test_cases.jsonnet#L19

  - IPU: [Colossus MK1 IPU](https://www.graphcore.ai/products/ipu)

>   - HPU: [AWS EC2 dl1.24xlarge](https://aws.amazon.com/ec2/instance-types/dl1/) which has 8 Gaudi accelerators

Reviewer: ```suggestion
  - HPU: [Intel Habana Gaudi SYS-420GH-TNGR](https://www.supermicro.com/en/products/system/AI/4U/SYS-420GH-TNGR) which has 8 Gaudi accelerators
```

## Documentation

| workflow file                      | action                                                                                       |
| ---------------------------------- | -------------------------------------------------------------------------------------------- |
| .github/workflows/ci_test-base.yml | Validate code examples in docstrings in the package with pytest’s doctest.                   |
| .github/workflows/docs-checks.yml  | Run doctest, build documentation, and upload built docs to make them available as artifacts. |
| .circleci/config.yml (build-docs)  | Build docs and host them on output.circleci-artifacts.com for easy access to the built docs. |
| .github/workflows/docs-link.yml    | Provide a direct link to built docs on output.circleci-artifacts.com.                        |

## Code Quality

| workflow file                     | action                                 |
| --------------------------------- | -------------------------------------- |

> | .codecov.yml                      | Measure test coverage with codecov.io. |

Reviewer: ```suggestion
| .codecov.yml                      | Measure test coverage with [codecov.io](https://app.codecov.io/gh/PyTorchLightning/pytorch-lightning) |
```

| .github/workflows/code-checks.yml | Check typing with mypy.                |
| .github/workflows/ci_schema.yml   | Validate the syntax of workflow files. |

## Others

| workflow file                          | action                                                                                                                                                                                                               |
| -------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | .github/workflows/ci_dockers.yml       | Build docker images used for testing in CI without pushing to the Docker Hub (pytorchlightning/pytorch_lightning). Publishing these built images takes place in .github/release-docker.yml which only runs in mater. |

Reviewer: ```suggestion
| .github/workflows/ci_dockers.yml       | Build docker images used for testing in CI without pushing to the [Docker Hub](https://hub.docker.com/r/pytorchlightning/pytorch_lightning). Publishing these built images takes place in `.github/release-docker.yml` which only runs in mater. |
```

<!-- Note: This document cannot be in `.github/README.md` because it will overwrite the repo README.md -->

# Continuous Integration and Delivery

## Unit and Integration Testing

| workflow name               | workflow file                       | action                                                                                                                                                                                      | accelerator\* | (Python, PyTorch)                                | OS                  |
| --------------------------- | ----------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------- | ------------------------------------------------ | ------------------- |

> | Test full                   | .github/workflows/ci_test-full.yml  | Run all tests except for accelerator-specific, standalone and slow tests.                                                                                                                   | CPU           | (3.7, 1.8), (3.7, 1.11), (3.9, 1.8), (3.9, 1.11) | linux, mac, windows |

Reviewer 3: yes we have a couple cpu-only horovod tests

> | Test full                   | .github/workflows/ci_test-full.yml  | Run all tests except for accelerator-specific, standalone and slow tests.                                                                                                                   | CPU           | (3.7, 1.8), (3.7, 1.11), (3.9, 1.8), (3.9, 1.11) | linux, mac, windows |

Author: @Borda How do you propose I should change?

> | Test full                   | .github/workflows/ci_test-full.yml  | Run all tests except for accelerator-specific, standalone and slow tests.                                                                                                                   | CPU           | (3.7, 1.8), (3.7, 1.11), (3.9, 1.8), (3.9, 1.11) | linux, mac, windows |

Reviewer: not sure, probably in a sense that some minimal CPU version is running here

> | Test full                   | .github/workflows/ci_test-full.yml  | Run all tests except for accelerator-specific, standalone and slow tests.                                                                                                                   | CPU           | (3.7, 1.8), (3.7, 1.11), (3.9, 1.8), (3.9, 1.11) | linux, mac, windows |

Author: @Borda I meant "all tests except for accelerator-specific" to be the test cases including CPU tests for horovod.

Would it be possible for you to directly change the statement if still needed?

| Test slow                   | .github/workflows/ci_test-slow.yml  | Run only slow tests. Slow tests usually need to spawn threads and cannot be speed up or simplified.                                                                                         | CPU           | (3.7, 1.8)                                       | linux, mac, windows |
| PL.pytorch-lightning (IPUs) | .azure-pipelines/ipu-tests.yml      | Run only IPU-specific tests.                                                                                                                                                                | IPU           | (3.8, 1.9)                                       | linux               |
| PL.pytorch-lightning (HPUs) | .azure-pipelines/hpu-tests.yml      | Run only HPU-specific tests.                                                                                                                                                                | HPU           | (3.8, 1.10)                                      | linux               |

> | PL.pytorch-lightning (GPUs) | .azure-pipelines/gpu-tests.yml      | Run GPU-specific tests as well as non-accelerator-specific, standalone and example tests. Each standalone test needs to run individually to avoid unwanted interactions between test cases. | GPU           | (3.7, 1.8)                                       | linux               |

Reviewer 3: CLU typo

> | PL.pytorch-lightning (GPUs) | .azure-pipelines/gpu-tests.yml      | Run GPU-specific tests as well as non-accelerator-specific, standalone and example tests. Each standalone test needs to run individually to avoid unwanted interactions between test cases. | GPU           | (3.7, 1.8)                                       | linux               |

Author: @Borda I'll apply the suggestion above!

> are we sure about this? we do not run any parallelization so what case conflict could happen... @tchaton could you pls help to clarify?

I may not get what you're concerned about with this description. I added it to describe that each test case marked `standalone=True` doesn't run in parallel with any other test cases. Any suggestion in particular?

> | PL.pytorch-lightning (GPUs) | .azure-pipelines/gpu-tests.yml      | Run GPU-specific tests as well as non-accelerator-specific, standalone and example tests. Each standalone test needs to run individually to avoid unwanted interactions between test cases. | GPU           | (3.7, 1.8)                                       | linux               |

Reviewer 4: I think "run individually" -> "run in separate processes" would clarify this confusion

> | PL.pytorch-lightning (GPUs) | .azure-pipelines/gpu-tests.yml      | Run GPU-specific tests as well as non-accelerator-specific, standalone and example tests. Each standalone test needs to run individually to avoid unwanted interactions between test cases. | GPU           | (3.7, 1.8)                                       | linux               |

Author: Done! Thank you for your suggestion.

| PyTorchLightning.Benchmark  | .azure-pipelines/gpu-benchmark.yml  | Run speed/memory benchmarks against pure PyTorch.                                                                                                                                           | GPU           | (3.7, 1.8)                                       | linux               |
| test-on-tpus                | .circleci/config.yml                | Run only TPU-specific tests.                                                                                                                                                                | TPU           | (3.7, 1.9)                                       | linux               |

- \*Accelerators used in CI
  - GPU: 2 x NVIDIA P100
  - TPU: Google GKE TPUv2/3
  - IPU: [Colossus MK1 IPU](https://www.graphcore.ai/products/ipu)

>   - HPU: [AWS EC2 dl1.24xlarge](https://aws.amazon.com/ec2/instance-types/dl1/) which has 8 Gaudi accelerators

Author: Not confident about the IPU and HPU setups 👀

Reviewer 2: HPU setup sounds right :)

## Documentation

| workflow file                      | action                                                                                       |
| ---------------------------------- | -------------------------------------------------------------------------------------------- |

> | .github/workflows/ci_test-base.yml | Validate code examples in docstrings in the package with pytest’s doctest.                   |
> | .github/workflows/docs-checks.yml  | Run doctest, build documentation, and upload built docs to make them available as artifact.  |
> | .circleci/config.yml (build-docs)  | Build docs and host them on output.circleci-artifacts.com for easy access to the built docs. |
> | .github/workflows/docs-link.yml    | Provide a direct link to built docs on output.circleci-artifacts.com.                        |

Reviewer 5: ```suggestion
| .github/workflows/ci_test-base.yml | Validate code examples in docstrings in the package with pytest’s doctest.                   |
| .github/workflows/docs-checks.yml  | Run doctest, build documentation, and upload built docs to make them available as artifacts.  |
| .circleci/config.yml (build-docs)  | Build docs and host them on output.circleci-artifacts.com for easy access to the built docs. |
| .github/workflows/docs-link.yml    | Provide a direct link to built docs on output.circleci-artifacts.com.                        |
```

| .github/workflows/legacy-checkpoints.yml | Generate legacy checkpoints and upload them to AWS S3.                             |
| .github/workflows/events-nightly.yml     | Publish the package to TestPyPI. Publish Docker images on hub.docker.com.          |

## Bots

| workflow file                                                      | action                                                                                                                                                                                                                    |
| ------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| .github/mergify.yml                                                | Label PRs as conflicts or ready, and request reviews if needed.                                                                                                                                                           |

> | .github/stale.yml                                                  | Close inactive issues/PRs sometimes after adding won't fix the label to them.                                                                                                                                             |

Author: ```suggestion
| .github/stale.yml                                                  | Close inactive issues/PRs sometimes after adding the "won't fix" label to them.                                                                                                                                             |
```

> | .github/mergify.yml                                                | Label PRs as conflicts or ready, and requests reviews if needed.                                                                                                                                                          |
> | .github/stale.yml                                                  | Close inactive issues/PRs some time after adding won't fix label to them.                                                                                                                                                 |
> | .github/workflows/probot-auto-cc.yml, .github/lightning-probot.yml | Notify maintainers of interest depending on labels added to an issue We utilise lightning-probot forked from PyTorch’s probot.                                                                                            |
> | .pre-commit-config.yaml                                            | pre-commit.ci runs a set of linters and formatters, such as black, flake8 and isort. When formatting is applied, the bot pushes a commit with its change. This configuration is also used for running pre-commit locally. |
> | .github/workflows/ci_pr-gatekeeper.yml                             | Prevent PRs from merging into master without any Grid.ai employees’ approval.                                                                                                                                             |
> 
> ## Dashboards
> 
> - CI status in master (GitHub Actions only) hosted by the PyTorch team: https://hud2.pytorch.org/ci/PyTorchLightning/pytorch-lightning/master/
> - Azure Pipelines (GPUs): https://dev.azure.com/PytorchLightning/pytorch-lightning/_pipeline/analytics/stageawareoutcome?definitionId=6&contextType=build

Reviewer 5: ```suggestion
| .github/mergify.yml                                                | Label PRs as conflicts or ready, and request reviews if needed.                                                                                                                                                          |
| .github/stale.yml                                                  | Close inactive issues/PRs sometimes after adding won't fix the label to them.                                                                                                                                                 |
| .github/workflows/probot-auto-cc.yml, .github/lightning-probot.yml | Notify maintainers of interest depending on labels added to an issue We utilize lightning-probot forked from PyTorch’s probot.                                                                                            |
| .pre-commit-config.yaml                                            | pre-commit.ci runs a set of linters and formatters, such as black, flake8 and isort. When formatting is applied, the bot pushes a commit with its change. This configuration is also used for running pre-commit locally. |
| .github/workflows/ci_pr-gatekeeper.yml                             | Prevent PRs from merging into master without any Grid.ai employees’ approval.                                                                                                                                             |

## Dashboards

- CI status in master (GitHub Actions only) hosted by the PyTorch team: https://hud2.pytorch.org/ci/PyTorchLightning/pytorch-lightning/master/
- Azure Pipelines (GPUs): https://dev.azure.com/PytorchLightning/pytorch-lightning/_pipeline/analytics/stageawareoutcome?definitionId=6&contextType=build
```
